Google.Api.Generator/Generation/MethodDetails.cs

```diff
@@ -275,6 +275,72 @@ private static MethodDetails DetectPagination(ServiceDetails svc, MethodDescript
                 field.IsRepeated && !field.IsMap && (field.FieldType == FieldType.Message || field.FieldType == FieldType.String);
         }
 
+        private static MethodDetails DetectRegapicPagination(ServiceDetails svc, MethodDescriptor desc)
+        {
+            var input = desc.InputType;
+            var output = desc.OutputType;
+
+            var resultsLimitCandidate = FindResultsLimitCandidate(input, IsResultsLimitCandidate);
+            var pageTokenCandidate = input.FindFieldByName("page_token");
+            var nextPageTokenCandidate = output.FindFieldByName("next_page_token");
+
+            var mapCandidateFields = output.Fields.InDeclarationOrder().Where(f => f.IsMap).ToList();
+            var repeatedCandidateFields = output.Fields.InDeclarationOrder().Where(IsRepeatedCandidate).ToList();
+            var repeatedField = mapCandidateFields.SingleOrDefault() ?? repeatedCandidateFields.SingleOrDefault();
+
+            if (resultsLimitCandidate != null && pageTokenCandidate != null && nextPageTokenCandidate != null && repeatedField != null)
+            {
+                if (!IsResultsLimitCandidate(resultsLimitCandidate))
+                {
+                    throw new InvalidOperationException($"{resultsLimitCandidate.Name} must be of type int32");
+                }
+
+                if (pageTokenCandidate.FieldType != FieldType.String || pageTokenCandidate.IsRepeated)
+                {
+                    throw new InvalidOperationException("page_token must be of type string");
+                }
+                if (nextPageTokenCandidate.FieldType != FieldType.String || nextPageTokenCandidate.IsRepeated)
+                {
+                    throw new InvalidOperationException("next_page_token must be of type string");
+                }
+                
+                return new Paginated(svc, desc, repeatedField, resultsLimitCandidate.FieldNumber, pageTokenCandidate.FieldNumber);
+            }
+            return null;
+
+            bool IsResultsLimitCandidate(FieldDescriptor field) =>
+                field is {FieldType: FieldType.Int32, IsRepeated: false};
+
+            bool IsRepeatedCandidate(FieldDescriptor field) =>
+                field.IsRepeated && !field.IsMap && (field.FieldType == FieldType.Message || field.FieldType == FieldType.String);
+        }
+
+        private static FieldDescriptor FindResultsLimitCandidate(MessageDescriptor input, Func<FieldDescriptor, bool> isResultsLimitCandidate)
```

Comment:
Rather than pass this as a delegate, could we just make `IsResultsLimitCandidate` a regular method and call it directly from this method? I think it would make things simpler. (I started looking to see whether we used different methods for this parameter.)

Reply:
👍